File: src/main/java/com/sofodev/thedragonlib/util/ItemStackHandlerImproved.java
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package com.sofodev.thedragonlib.util;

import net.minecraft.core.NonNullList;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.items.ItemStackHandler;

import javax.annotation.Nonnull;

public class ItemStackHandlerImproved extends ItemStackHandler {

    private boolean tempIgnoreConditions;

    public ItemStackHandlerImproved(int slots) {
        super(slots);
    }

    public NonNullList<ItemStack> getItems() {
        return this.stacks;
    }

    @Nonnull
    @Override
    public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate) {
        return !this.tempIgnoreConditions && !this.canInsert(stack, slot) ? stack : super.insertItem(slot, stack, simulate);
    }

    @Nonnull
    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        return !this.tempIgnoreConditions && !this.canExtract(this.getStackInSlot(slot), slot) ? ItemStackUtils.getNull() : super.extractItem(slot, amount, simulate);
    }

    public boolean canInsert(ItemStack stack, int slot) {
        return true;
    }

    public boolean canExtract(ItemStack stack, int slot) {
        return true;
    }
}

File: src/main/java/net/thedragonteam/thedragonlib/util/ItemStackHandlerImproved.kt
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.thedragonlib.util

import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraftforge.items.ItemStackHandler

class ItemStackHandlerImproved(slots: Int) : ItemStackHandler(slots) {

    private val tempIgnoreConditions: Boolean = false

    val items: NonNullList<ItemStack>
        get() = this.stacks

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        return if (!this.tempIgnoreConditions && !this.canInsert(stack, slot)) stack else super.insertItem(slot, stack, simulate)
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        return if (!this.tempIgnoreConditions && !this.canExtract(this.getStackInSlot(slot), slot)) ItemStackUtils.getNull() else super.extractItem(slot, amount, simulate)
    }

    fun canInsert(stack: ItemStack, slot: Int): Boolean {
        return true
    }

    fun canExtract(stack: ItemStack, slot: Int): Boolean {
        return true
    }
}